src/components/canvas-tools/text-tool.tsx

```diff
@@ -0,0 +1,95 @@
+import * as React from "react";
+import memoize from "memoize-one";
+import { Editor } from "slate-react";
+import Plain from "slate-plain-serializer";
+import { ToolTileModelType } from "../../models/tools/tool-tile";
+import { TextContentModelType } from "../../models/tools/text/text-content";
+import { Value } from "slate";
+
+import "./text-tool.sass";
+​
+interface IProps {
+  model: ToolTileModelType;
+  readOnly?: boolean;
+}
+
+interface IState {
```

Comment:
I don't think we should be using local state but instead keep it all in the MST model so it is more easily tested.

Reply:
You may be right, but that was a conscious decision. Serializing to Firebase is expensive and time-consuming and I was thinking that doing so on every keystroke would be excessive while doing so on blur would be sufficient, but you're right that some things would be simpler that way so maybe we should start there and see how it goes.

Reply:
If we used ProseMirror instead of Slate it supports transactions which we could just push to firebase instead of storing all the text on each change.  Here is an example project that does this: https://github.com/xylk/prosemirror-firebase

Reply:
Dave and I looked at ProseMirror (and Quill) as well. There were a couple things we liked better about Slate, including the built-in React integration, but I'd consider the work so far to be a spike and better firebase integration may well be a point in ProseMirror's favor.